### PR TITLE
Refactor/deployments

### PR DESCRIPTION
**Pull request by 0xlucian**
## Description

This PR refactors the deployments scripts and adds a new config file. In this new config file all of the tokens and pools parameters will be defined and deployment scripts will consume this config rather than having hardcoded parameters.

Resolves # VEN-686

## Checklist

- [ ] I have updated the documentation to account for the changes in the code.
- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a test preventing this bug from silently reappearing again.
- [ ] My contribution follows [Venus contribution guidelines](docs/CONTRIBUTING.md).

### CODE REVIEW

File: deploy/003-price-oracle-configure-feeds.ts [hunk 1]
```diff
@@ -9,6 +9,6 @@ const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
 };
 
 func.tags = ["Oracle"];
-func.skip = async (hre: HardhatRuntimeEnvironment) => hre.network.name !== "hardhat";
+func.skip = async (hre: HardhatRuntimeEnvironment) => hre.network.live;
```
Comment (coreyar) [inline deploy/003-price-oracle-configure-feeds.ts:12]: We need to remove line 7

Comment (0xlucian) [reply deploy/003-price-oracle-configure-feeds.ts]: We don't actually need to define and extend `DeployFunction` as it is defined and the logic is in the deployment we import from... pushed the fix for that

File: contracts/Pool/PoolRegistry.sol [hunk 2]
```diff
@@ -249,7 +249,7 @@ contract PoolRegistry is Ownable2StepUpgradeable {
             input.asset,
             comptroller,
             rate,
-            10**input.decimals,
+            10**(input.decimals - 8 + 18),
```
Comment (chechu) [inline contracts/Pool/PoolRegistry.sol:252]: I think `input.decimals` is the number of decimals of the VToken: 8, as you can see in the script number 10 [here](https://github.com/VenusProtocol/isolated-pools/pull/138/files#diff-402568a4acf31df6c7c204e6fd264f2985f0e229ffe0d2ecf5d7e6eded4de8dbR108), but what we need here is the number de decimals of the underlying token.

If we assume the underlying token is a ERC20 token we might get the number of decimal places in a dynamic way, invoking `decimals()` on the provided address. What do you think?

File: deploy/006-riskfund-protocolshare.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    log: true,
+  });
+
+  await riskFund.setShortfallContractAddress(shortfallDeployment.address);
+
+  await deploy("ProtocolShareReserve", {
+    from: deployer,
+    log: true,
+    autoMine: true,
+  });
```
[the remark this answers is not in the document]
Comment (0xlucian) [reply deploy/006-riskfund-protocolshare.ts:69]: Yeah we need this contract to be deployed

File: deploy/008-access-control-configure.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import { DeployFunction } from "hardhat-deploy/types";
+import { HardhatRuntimeEnvironment } from "hardhat/types";
+
+const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
+  const { deployments, getNamedAccounts } = hre;
+  const { deploy } = deployments;
+  const { deployer } = await getNamedAccounts();
+
+  const proxyAdmin = await ethers.getContract("DefaultProxyAdmin");
+  console.log("Proxy Admin: " + proxyAdmin.address);
```
[the remark this answers is not in the document]
Comment (0xlucian) [reply deploy/008-access-control-configure.ts:11]: Nope I can remove this.. I wanted do make sure we are getting the right ProxyAdmin

File: deploy/010-deploy-pools.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    for (const vtoken of pool.vtokens) {
+      const token = getTokenConfig(vtoken.asset, tokenConfig);
+      let tokenContract;
+      if (token.isMock) {
+        tokenContract = await ethers.getContract(`Mock${token.symbol}`);
+        console.log("Minting " + vtoken.initialSupply + " mock tokens to owner");
+        await tokenContract.faucet(vtoken.initialSupply);
+      } else {
+        tokenContract = await ethers.getContractAt("ERC20", token.tokenAddress);
+        // Make sure that deployer has at least `initialSupply` balance of the token
```
[the remark this answers is not in the document]
Comment (0xlucian) [reply deploy/010-deploy-pools.ts:82]: Nope.. If we make sure the deployer has enough token balance, and `tokenAddress` is correct we should be all fine.

File: contracts/Pool/PoolRegistry.sol (hunk 2 above)
Comment (0xlucian) [reply contracts/Pool/PoolRegistry.sol:252]: You are right... What bothers me is why tests pass 🤔

File: deploy/006-riskfund-protocolshare.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (chechu) [reply deploy/006-riskfund-protocolshare.ts:69]: But we are deploying the same contract two times

[the remark this answers is not in the document]
Comment (0xlucian) [reply deploy/006-riskfund-protocolshare.ts]: You are right. This is fixed

File: contracts/Pool/PoolRegistry.sol (hunk 2 above)
Comment (0xlucian) [reply contracts/Pool/PoolRegistry.sol]: Already fixed

File: tests/integration/index.ts [hunk 6]
```diff
@@ -186,81 +214,79 @@ describe("Positive Cases", () => {
...
       expect(borrowBalance).to.equal(0);
       [error, liquidity, shortfall] = await Comptroller.connect(acc2Signer).getAccountLiquidity(acc2);
       expect(error).to.equal(Error.NO_ERROR);
-      expect(liquidity).to.equal(new BigNumber(mintAmount * collateralFactor).multipliedBy(vBNXPrice));
+      expect(liquidity).to.equal(new BigNumber(mintAmount).multipliedBy(collateralFactor).multipliedBy(vBNXPrice));
       expect(shortfall).to.equal(0);
       // ////////////
       // // Borrow //
       // ////////////
       // Supply WBTC to market from 2nd account
```
Comment (chechu) [inline tests/integration/index.ts:230]:
```suggestion
      // Supply BSW to market from 1st account
```

File: tests/integration/index.ts [hunk 7]
```diff
@@ -186,81 +214,79 @@ describe("Positive Cases", () => {
...
       await expect(vBSW.connect(acc1Signer).mint(mintAmount))
         .to.emit(vBSW, "Mint")
-        .withArgs(await acc1Signer.getAddress(), mintAmount, mintAmount, mintAmount);
-
+        .withArgs(await acc1Signer.getAddress(), mintAmount, vTokenMintAmount, vTokenMintAmount);
       [error, balance, borrowBalance] = await vBSW
         .connect(acc2Signer)
```
Comment (chechu) [inline tests/integration/index.ts:235]:
```suggestion
        .connect(acc1Signer)
```

File: tests/integration/index.ts [hunk 8]
```diff
@@ -186,81 +214,79 @@ describe("Positive Cases", () => {
...
-
       [error, liquidity, shortfall] = await Comptroller.connect(acc1Signer).getAccountLiquidity(acc1);
       expect(error).to.equal(Error.NO_ERROR);
-      expect(liquidity).to.equal(new BigNumber(mintAmount * collateralFactor).multipliedBy(vBSWPrice));
+      expect(liquidity).to.equal(new BigNumber(mintAmount).multipliedBy(collateralFactor).multipliedBy(vBSWPrice));
       expect(shortfall).to.equal(0);
-
-      const bswBorrowAmount = 1e4;
+      const bswBorrowAmount = convertToUnit(1, 18);
+      const vBSWBorrowAmount = convertToUnit(1, 18);
```
Comment (chechu) [inline tests/integration/index.ts:245]: I would say we don't need this variable. In the borrow flow we don't deal with vTokens

File: tests/integration/index.ts [hunk 9]
```diff
@@ -186,81 +214,79 @@ describe("Positive Cases", () => {
...
 
       await expect(vBSW.connect(acc2Signer).borrow(bswBorrowAmount))
         .to.emit(vBSW, "Borrow")
-        .withArgs(acc2, bswBorrowAmount, bswBorrowAmount, bswBorrowAmount);
-
+        .withArgs(acc2, bswBorrowAmount, vBSWBorrowAmount, vBSWBorrowAmount);
```
Comment (chechu) [inline tests/integration/index.ts:249]: The third parameter is `accountBorrowsNew` and the last one `totalBorrowsNew`, so in both cases they are referencing the underlying token.
```suggestion
        .withArgs(acc2, bswBorrowAmount, bswBorrowAmount, bswBorrowAmount);
```

File: tests/integration/index.ts [hunk 10]
```diff
@@ -186,81 +214,79 @@ describe("Positive Cases", () => {
...
       [error, balance, borrowBalance] = await vBSW.connect(acc2Signer).getAccountSnapshot(acc2);
       expect(error).to.equal(Error.NO_ERROR);
       expect(balance).to.equal(0);
-      expect(borrowBalance).to.equal(bswBorrowAmount);
+      expect(borrowBalance).to.equal(vBSWBorrowAmount);
```
Comment (chechu) [inline tests/integration/index.ts:253]:
```suggestion
      expect(borrowBalance).to.equal(bswBorrowAmount);
```

File: tests/integration/index.ts [hunk 11]
```diff
@@ -526,40 +570,38 @@ describe("Straight Cases For Single User Liquidation and healing", () => {
...
+      await vBNX.connect(acc2Signer).mint(mintAmount);
       // heal
       await expect(Comptroller.connect(acc1Signer).healAccount(acc2))
         .to.be.revertedWithCustomError(Comptroller, "CollateralExceedsThreshold")
-        .withArgs("100000000000000000000", "159990000015999000000");
+        .withArgs("100000000000000000000", "159990159990000000000");
     });
 
     it("Should revert on healing if borrow is less then collateral amount", async function () {
+      // await vBNX.connect(acc2Signer).redeem(convertToUnit("5", 7));
```
Comment (chechu) [inline tests/integration/index.ts:583]:
```suggestion
```

File: tests/integration/index.ts (hunk 10 above)
Comment (0xlucian) [reply tests/integration/index.ts:253]: BorrowBalance is with the token precision. We should not use the vTokenBorrow amount but the underlying token amount

File: tests/integration/index.ts (hunk 9 above)
Comment (0xlucian) [reply tests/integration/index.ts:249]: Same here

File: tests/integration/index.ts (hunk 7 above)
Comment (0xlucian) [reply tests/integration/index.ts:235]: We are supplying with acc2 to the market so acc1 to be able to borrow.

File: tests/integration/index.ts (hunk 10 above)
Comment (0xlucian) [reply tests/integration/index.ts]: This is fixed

File: tests/integration/index.ts (hunk 9 above)
Comment (0xlucian) [reply tests/integration/index.ts]: Fixed

File: tests/integration/index.ts (hunk 8 above)
Comment (0xlucian) [reply tests/integration/index.ts]: Removed it.

File: tests/integration/index.ts (hunk 7 above)
Comment (0xlucian) [reply tests/integration/index.ts]: This is fixed

**Review by chechu — APPROVED** [review]

**Pull request merged**